internal/credential/rotator.go

```diff
@@ -0,0 +1,254 @@
+	r.primaryEnvironmentID = envID
+	r.additions = append(r.additions, envID)
+	if previous.Defined() {
+		r.loggers.Infof("Environment ID %s was rotated, new environment ID is %s", r.primaryEnvironmentID, envID)
+		r.expirations = append(r.expirations, previous)
+	} else {
+		r.loggers.Infof("New environment ID is %s", envID)
+	}
+}
+
+func (r *Rotator) updateMobileKey(mobileKey config.MobileKey) {
+	if mobileKey == r.MobileKey() {
+		return
+	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	previous := r.primaryMobileKey
+	r.primaryMobileKey = mobileKey
+	r.additions = append(r.additions, mobileKey)
+	if previous.Defined() {
+		r.expirations = append(r.expirations, previous)
+		r.loggers.Infof("Mobile key %s was rotated, new primary mobile key is %s", previous.Masked(), mobileKey.Masked())
+	} else {
+		r.loggers.Infof("New primary mobile key is %s", mobileKey.Masked())
+	}
+}
+
+func (r *Rotator) swapPrimaryKey(newKey config.SDKKey) config.SDKKey {
+	if newKey == r.primarySdkKey {
+		// There's no swap to be done, we already are using this as primary.
+		return ""
+	}
+	previous := r.primarySdkKey
+	r.primarySdkKey = newKey
+	r.additions = append(r.additions, newKey)
+	r.loggers.Infof("New primary SDK key is %s", newKey.Masked())
+
+	return previous
+}
+func (r *Rotator) updateSDKKey(sdkKey config.SDKKey, grace *GracePeriod) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	previous := r.swapPrimaryKey(sdkKey)
+	// Immediately revoke the previous SDK key if there's no explicit deprecation notice, otherwise it would
+	// hang around forever.
+	if previous.Defined() && grace == nil {
+		r.expirations = append(r.expirations, previous)
+		r.loggers.Infof("SDK key %s has been immediately revoked", previous.Masked())
+		return
+	}
+	if grace != nil {
+		if previousExpiry, ok := r.deprecatedSdkKeys[grace.key]; ok {
+			if previousExpiry != grace.expiry {
+				r.loggers.Warnf("SDK key %s was marked for deprecation with an expiry at %v, but it was previously deprecated with an expiry at %v. The previous expiry will be used. ", grace.key.Masked(), grace.expiry, previousExpiry)
+			}
+			return
+		}
+
+		if grace.now.After(grace.expiry) {
+			r.loggers.Infof("Deprecated SDK key %s already expired; ignoring", grace.key.Masked())
+			return
+		}
+
+		r.loggers.Infof("SDK key %s was marked for deprecation with an expiry at %v", grace.key.Masked(), grace.expiry)
+		r.deprecatedSdkKeys[grace.key] = grace.expiry
+
+		if grace.key != previous {
+			r.loggers.Infof("Deprecated SDK key %s was not previously managed by Relay", grace.key.Masked())
+			r.additions = append(r.additions, grace.key)
+		}
+	}
+}
+
+func (r *Rotator) expireSDKKey(sdkKey config.SDKKey) {
+	r.loggers.Infof("Deprecated SDK key %s has expired and is no longer valid for authentication", sdkKey.Masked())
+	delete(r.deprecatedSdkKeys, sdkKey)
+	r.expirations = append(r.expirations, sdkKey)
+}
+
+func (r *Rotator) Query(now time.Time) (additions []SDKCredential, expirations []SDKCredential) {
```

Comment:
nit: think of a better name for this since query implies nothing is changed internally (idempotent)